spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,288 @@
+    }
+
+    Preconditions.checkState(cachedPendingBatches != null,
+        "pendingBatches is null, which is unexpected as it will be set when calculating end offset");
+
+    List<MicroBatch> pendingBatches = cachedPendingBatches.second();
+    if (pendingBatches.isEmpty()) {
+      LOG.info("There's no task to process in this batch");
+      return Collections.emptyList();
+    }
+
+    MicroBatch lastBatch = pendingBatches.get(pendingBatches.size() - 1);
+    if (lastBatch.snapshotId() != endOffset.snapshotId() || lastBatch.endFileIndex() != endOffset.index()) {
+      throw new IllegalStateException("The cached pendingBatches doesn't match the current end offset " + endOffset);
+    }
+
+    LOG.info("Processing data from {} to {}", startOffset, endOffset);
+    List<FileScanTask> tasks = pendingBatches.stream()
+        .flatMap(batch -> batch.tasks().stream())
+        .collect(Collectors.toList());
+    CloseableIterable<FileScanTask> splitTasks = TableScanUtil.splitFiles(CloseableIterable.withNoopClose(tasks),
+        splitSize());
+    return Lists.newArrayList(
+        TableScanUtil.planTasks(splitTasks, splitSize(), splitLookback(), splitOpenFileCost()));
+  }
+
+  private StreamingOffset calculateStartingOffset() {
+    StreamingOffset startingOffset;
+    if (startSnapshotId != null) {
+      startingOffset = new StreamingOffset(startSnapshotId, 0, true, false);
+    } else {
+      List<Long> snapshotIds = SnapshotUtil.currentAncestors(table);
+      if (snapshotIds.isEmpty()) {
+        // there's no snapshot currently.
+        startingOffset = StreamingOffset.START_OFFSET;
+      } else {
+        startingOffset = new StreamingOffset(snapshotIds.get(snapshotIds.size() - 1), 0, true, false);
+      }
+    }
+
+    return startingOffset;
+  }
+
+  private StreamingOffset calculateEndOffset(StreamingOffset start) {
+    if (start.equals(StreamingOffset.START_OFFSET)) {
+      return StreamingOffset.START_OFFSET;
+    }
+
+    // Spark will invoke setOffsetRange more than once. If this is already calulated, use the cached one to avoid
+    // calculating again.
+    if (cachedPendingBatches == null || !cachedPendingBatches.first().equals(start)) {
+      this.cachedPendingBatches = Pair.of(start, getChangesWithRateLimit(start.snapshotId(), start.index(),
+          start.isStartingSnapshotId(), start.isLastIndexOfSnapshot(), maxSizePerBatch));
+    }
+
+    List<MicroBatch> batches = cachedPendingBatches.second();
+    MicroBatch lastBatch = batches.isEmpty() ? null : batches.get(batches.size() - 1);
+
+    if (lastBatch == null) {
+      return start;
+    } else {
+      boolean isStarting = lastBatch.snapshotId() == start.snapshotId() && start.isStartingSnapshotId();
+      return new StreamingOffset(lastBatch.snapshotId(), lastBatch.endFileIndex(), isStarting,
+          lastBatch.lastIndexOfSnapshot());
+    }
+  }
+
+  @VisibleForTesting
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```

Comment:
I don't think that this should suppress the complexity check. I agree that this method is a bit long and difficult to read.

It would help to rewrite it so that the loop from the given snapshotId to the table's current snapshotId handles the snapshotId passed in as well. That would make this much easier to read.